### PR TITLE
XWIKI-14548: Add exclusive notifications filters

### PR DESCRIPTION
**Pull request by aubincleme**
* Add NotificationPreferenceScopeFilterType enum
* Add support for `#queryFilterAnd()` in `ScopeNotificationFilter`
* Add unit tests

### CODE REVIEW

**Comment by aubincleme** [comment]
Here is a screenshot of what the UI looks like when adding a new filter:

selection_002

File: xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java [hunk 1]
```diff
@@ -66,18 +66,64 @@
...
         boolean hasRestriction = false;
         // Indicate if a restriction matches the document of the event
         boolean matchRestriction = false;
 
         try {
-            for (NotificationPreferenceScope scope : modelBridge.getNotificationPreferenceScopes(user, format)) {
+            for (NotificationPreferenceScope scope : modelBridge.getNotificationPreferenceScopes(user, format,
+                    scopeFilterType)) {
                 if (scope.getEventType().equals(event.getType())) {
                     hasRestriction = true;
```
Comment (gdelhumeau) [inline xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java]: I think hasRestriction should be true only with NotificationPreferenceScopeFilterType.INCLUSIVE

File: xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java [hunk 2]
```diff
@@ -87,113 +133,142 @@ public boolean filterEvent(Event event, DocumentReference user, NotificationForm
             logger.warn(ERROR, e);
         }
 
-        return hasRestriction && !matchRestriction;
+        /**
+         * In case we have an INCLUSIVE filter, we check if we had a restriction that was not satisfied.
+         * In the case of an EXCLUSIVE filter, if a restriction has been found, then the function should have already
+         * returned true.
+         */
+        if (scopeFilterType.equals(NotificationPreferenceScopeFilterType.INCLUSIVE)) {
```
Comment (gdelhumeau) [inline xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java]: why not:
```
return scopeFilterType.equals(NotificationPreferenceScopeFilterType.INCLUSIVE) && hasRestriction && !matchRestriction;
```

File: xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java (hunk 1 above)
Comment (aubincleme) [inline xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java]: If we’re working with `NotificationPreferenceScopeFilterType.EXCLUSIVE`, we don’t actually care about the value of `hasRestriction`.

File: xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java (hunk 2 above)
Comment (aubincleme) [inline xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java]: That’s a good idea ; fixing it.

Comment (aubincleme) [inline xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java]: Fixed in 95b72c1fdb.

**Pull request merged**